File: RestoreIPAddresses.py
"""
93. Restore IP Addresses
Given a string containing only digits, restore it by returning all possible valid IP address combinations.

For example:
Given "25525511135",

return ["255.255.11.135", "255.255.111.35"]. (Order does not matter)
"""
# O(1) (O(3 ^ 4))
# O(1) (O(3 * 4))
class Solution(object):
    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        res = []
        self.recur(res, s, 0, len(s), [])
        return res
        
    def recur(self, res, s, start, end, cur):
        if len(cur) == 4:
            if start == end:
                res.append(".".join(cur))
            return
        if start >= end:
            return
        if s[start] == "0":
            cur.append("0")
            self.recur(res, s, start + 1, end, cur)
            cur.pop()
            return
        for i in range(3):
            tem = s[start : start + i + 1]
            if int(tem) <= 255:
                cur.append(tem)
                self.recur(res, s, start + i + 1, end, cur)
                cur.pop()

if __name__ == "__main__":
    print(Solution().restoreIpAddresses("010010"))

File: Java/RestoreIPAddresses.java
import java.util.*;
public class RestoreIPAddresses {
    public List<String> restoreIpAddresses(String s) {
        List<String> res = new ArrayList<String>();
        if(s.length() == 0) {return res;}
        List<String> cur = new ArrayList<String>();
        recur(res, cur, s, 0);
        return res;
    }
    
    private void recur(List<String> res, List<String> cur, String s, int start) {
        if(start >= s.length()) {
            if(cur.size() == 4) {
                res.add(String.join(".", cur));
            }
            return;
        }
        if(cur.size() == 4) {return;}
        if(s.charAt(start) == '0') {
            cur.add("0");
            recur(res, cur, s, start + 1);
            cur.remove(cur.size() - 1);
            return;
        }
        int tem = 0;
        for(int i = start; i < s.length(); i++) {
            tem = tem * 10 + s.charAt(i) - '0';
            if(tem <= 255) {
                cur.add(Integer.toString(tem));
                recur(res, cur, s, i + 1);
                cur.remove(cur.size() - 1);
            } else {
                break;
            }
        }
    }

	public static void main(String[] args) {
		String s = "12345";
		Character c;
		int a = 0;
		for(int i = 0; i < s.length(); i++) {
			c = s.charAt(i);
			a = a * 10 + s.charAt(i) - '0';
		}
		System.out.println(a);
		System.out.println(new RestoreIPAddresses().restoreIpAddresses("25525511135"));
		System.out.println(new RestoreIPAddresses().restoreIpAddresses("010010"));

	}

}
